packages/react-aria-components/src/Tree.tsx

```diff
@@ -0,0 +1,548 @@
+  // Whether the tree row is expanded.
+  isExpanded: boolean
+}
+
+export interface TreeItemProps<T = object> extends StyleRenderProps<TreeItemRenderProps>, LinkDOMProps {
+  /** The unique id of the tree row. */
+  id?: Key,
+  /** The object value that this tree row represents. When using dynamic collections, this is set automatically. */
+  value?: T,
+  /** A string representation of the tree row's contents, used for features like typeahead. */
+  textValue: string,
+  /** An accessibility label for this tree row. */
+  'aria-label'?: string,
+  /** A list of child tree row objects used when dynamically rendering the tree row children. */
+  childItems?: Iterable<T>,
+  // TODO: made this required since the user needs to pass Content at least
+  /** The content of the tree row along with any nested children. Supports static items or a function for dynamic rendering. */
+  children: ReactNode | ((item: T) => ReactElement)
+}
+
+function TreeItem<T extends object>(props: TreeItemProps<T>, ref: ForwardedRef<HTMLDivElement>): JSX.Element | null {
+  let {childItems, children} = props;
+  let render = useContext(CollectionRendererContext);
+  let childRows: ReactNode | ((item: T) => ReactNode);
+  let rowContent: ReactNode | null;
+
+  if (typeof render === 'function') {
+    childRows = render;
+    // Assumption here is that props.children[0] is Content
+    rowContent = children![0];
+  } else if (typeof children !== 'function') {
+    childRows = children;
+  }
+
+  let collectionChildren = useCollectionChildren({
+    children: childRows,
+    items: childItems
+  });
+
+  // Combine the renderChildren and children so both Content and nested TreeItems are properly added to fake DOM and thus added to the built collection
+  return useSSRCollectionNode('item', props, ref, null, [rowContent, collectionChildren]);
+}
+
+/**
+ * A TreeItem represents an individual item in a Tree.
+ */
+const _TreeItem = /*#__PURE__*/ (forwardRef as forwardRefType)(TreeItem);
+export {_TreeItem as TreeItem};
+
+export interface TreeItemContentRenderProps extends ItemRenderProps {
+  // Whether the tree row is expanded.
+  isExpanded: boolean,
+  // Whether the tree row has child rows.
+  hasChildRows: boolean,
+  // What level the tree row has within the tree.
+  level: number
+}
+
+// The TreeItemContent is the one that accepts RenderProps because we would get much more complicated logic in TreeItem otherwise since we'd
+// need to do a bunch of check to figure out what is the Content and what are the actual collection elements (aka child rows) of the TreeItem
+export interface TreeItemContentProps extends Pick<RenderProps<TreeItemContentRenderProps>, 'children'> {}
+
+// TODO does this need ref or context? Its only used to shallowly render the Content node... If it was a more generic collection component then I could see an argument for it
+// having those
+function TreeItemContent(props: TreeItemContentProps) {
```

Comment:
omitted ref here for the reason above, don't think it is very useful to have a ref + context unlike the generic Heading/Separator

Reply:
yep thats right